src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+| Address | Name | Value  |
++---------+------+--------+
+| 2 ^ 30  |      | 5      |
+| ...     | ...  | ...    |
+| 1       | y    | 42     |
+| 0       | x    | 2 ^ 30 |
+
+We have `2 ^ 30` in our hypothetical computer with 1GB of RAM. And since our
+stack grows from zero, the easiest place to allocate memory is from the other
+end. So our first value is at the highest place in memory. And the value of the
+struct at `x` has a [raw pointer][rawpointer] to the place we’ve allocated on
+the heap, so the value of `x` is `2 ^ 30`, the memory location we’ve asked for.
+
+[rawpointer]: raw-pointers.html
+
+Since this memory is on the heap, it can stay alive longer than the function
+which allocates the box. In this case, however, it doesn’t. When the function
+is over, we need to free the stack frame for `main()`. `Box<T>`, though, has a
+trick up its sleve: [Drop][drop]. The implementation of `Drop` for `Box`
+deallocates the memory that was allocated when it was created. Great! So when
+`x` goes away, it first frees the memory allocated on the heap:
+
+[drop]: drop.html
+
+| Address | Name | Value  |
++---------+------+--------+
+| 1       | y    | 42     |
+| 0       | x    | ?????? |
+
+And then the stack frame goes away, freeing all of our memory.
+
+# Arguments and borrowing
+
+We’ve got some basic examples with the stack and the heap going, but what about
+function arguments and borrowing? Here’s a small Rust program:
+
+```rust
+fn foo(i: &i32) {
+    let z = 42;
+}
+
+fn main() {
+    let x = 5;
+    let y = &x;
+
+    foo(y);
+}
+```
+
+When we enter `main()`, memory looks like this:
+
+| Address | Name | Value |
++---------+------+-------+
+| 1       | y    | 0     |
+| 0       | x    | 5     |
+
+`x` is a plain old `5`, and `y` is a reference to `x`. So its value is the
+memory location that `x` lives at, which in this case is `0`.
+
+What about when we call `foo()`, passing `y` as an argument?
+
+| Address | Name | Value |
++---------+------+-------+
+| 3       | z    | 42    |
+| 2       | i    | 0     |
+| 1       | y    | 0     |
+| 0       | x    | 5     |
+
+Stack frames aren’t just for local bindings, they’re for arguments too. So in
+this case, we need to have both `i`, our argument, and `z`, our local variable
+binding. `i` is a copy of the argument, `y`. Since `y`’s value is `0`, so is
+`i`’s.
+
+This is one reason why borrowing a variable doesn’t deallocate any memory: the
+value of a reference is just a pointer to a memory location. If we got rid of
+the underlying memory, things wouldn’t work very well.
+
+# A complex example
+
+Okay, let’s go through this complex program step-by-step:
+
+```rust
+fn foo(x: &i32) {
+    let y = 10;
+    let z = &y;
+
+    baz(z);
+    bar(x, z);
+}
+
+fn bar(a: &i32, b: &i32) {
+    let c = 5;
+    let d = Box::new(5);
+    let e = &d;
+
+    baz(e);
+}
+
+fn baz(f: &i32) {
+    let g = 100;
+}
+
+fn main() {
+    let h = 3;
+    let i = Box::new(20);
+    let j = &h;
+
+    foo(j);
+}
+```
```

Comment:
I like this, and want to improve this example, but yeah, I think this would work better as a new section, at the end. I think I'll file this under 'future improvements', as I want to make sure to tackle other doc stuff this week too.